[Feature] Support Dictionary (#35559)

srlch opened the pull request:
Why I'm doing:
StarRocks support `dict_mapping` function(#20961) for query a dictionary table using expression.
But there are still some limitations:
1. Only support primary key table.
2. Dictionary table lookups still require io and network overhead.
3. Multiple columns returns are not supported

What I'm doing:
StarRocks introduces `dictionary` objects. The dictionary object is similar to a memory table organized
into a hash table. The dictionary object caches all the data of the dictionary table to each be node,
and implements efficient dictionary queries in memory through the `dictionary_get` function.
Dictionary ojbect support all kinds of table modes as well as VIEW and MV.
Dictionary oject return multiple columns using struct type.

Usage:

MySQL [test]> SELECT * FROM t;
+------+------+------+
| id   | name | job1 |
+------+------+------+
|    1 |    2 |    3  |
+------+------+------+
1 row in set (0.01 sec)

MySQL [test]> CREATE DICTIONARY my_dictionary USING t (id KEY, name VALUE, job1 VALUE);
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> REFRESH DICTIONARY my_dictionary;
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> SELECT dictionary_get("my_dictionary", id) FROM t;
+-------------------------------------+
| dictionary_get('my_dictionary', id) |
+-------------------------------------+
| {"name":2,"job1":3}                 |
+-------------------------------------+
1 row in set (0.02 sec)

Fixes #35559

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

imay commented on gensrc/proto/internal_service.proto:
```diff
@@ -267,6 +267,19 @@ message PTabletWriterAddSegmentResult {
 };
 
+message PRefreshDictionaryCacheRequest {
+    required ChunkPB chunk = 1;
+    required int64 dictionary_id = 2;
+    required int64 txn_id = 3;
+    required POlapTableSchemaParam schema = 4;
+    required int64 memory_limit = 5;
+    required int32 key_size = 6;
```
please don't use `required` in thrift and proto

srlch replied:
done

imay commented on gensrc/thrift/DataSinks.thrift:
```diff
@@ -186,6 +187,15 @@ struct TExportSink {
+    3: required i64 txn_id
+    4: required Descriptors.TOlapTableSchemaParam schema
+    5: required i64 memory_limit
+    6: required i32 key_size
```
no required

srlch replied:
done

imay commented on gensrc/thrift/Exprs.thrift:
```diff
@@ -181,6 +184,12 @@ struct TDictQueryExpr {
+struct TDictionaryExpr {
+  1: required i64 dict_id
+  2: required i64 txn_id
+  3: required i32 key_size
```
no requried

srlch replied:
done

imay reviewed (changes requested):
This PR is too big, you need to split it into several smaller CL. It is hard to review so huge PR.

srlch commented:
> This PR is too big, you need to split it into several smaller CL. It is hard to review so huge PR.

I have split it into five smaller commit, PTAL

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/catalog/Dictionary.java:
```diff
@@ -0,0 +1,321 @@
+    private DictionaryState state = DictionaryState.UNINITIALIZED;
+    private DictionaryState stateBeforeRefresh = null;
+    private String runtimeErrMsg;
+
```
You should add empty constructor 
```
public Dictionary() {}
```
The default value will be null if there is no empty constructor when deserializing.

srlch replied:
done

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:
```diff
@@ -2030,6 +2044,8 @@ public void saveImage(File curFile, long replayedJournalId) throws IOException {
                 dos.writeLong(checksum);
                 checksum = pipeManager.getRepo().saveImage(dos, checksum);
                 dos.writeLong(checksum);
+                checksum = dictionaryMgr.saveDictionaryMgr(dos, checksum);
```
no need to save the old format

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/persist/DictionaryMgrInfo.java:
```diff
@@ -0,0 +1,73 @@
+    private long lastSuccessTxnId = -1L;
+    @SerializedName(value = "nextDictionaryId")
+    private long nextDictionaryId = 1L;
+
```
Same to Dictionary.java, should add empty constructor

srlch replied:
done

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/persist/DropDictionaryInfo.java:
```diff
@@ -0,0 +1,48 @@
+public class DropDictionaryInfo implements Writable {
+    @SerializedName(value = "dictionaryName")
+    String dictionaryName;
+
```
Add empty contructor

srlch replied:
done

srlch commented on fe/fe-core/src/main/java/com/starrocks/catalog/Dictionary.java:
```diff
@@ -0,0 +1,321 @@
+
+public class Dictionary implements Writable, GsonPostProcessable {
+    // Dictionary properties
+    public static final String PROPERTIES_DICTIONARY_WARMING_UP = "dictionary_warming_up";
```
done

srlch commented on be/src/storage/dictionary_cache_manager.h:
```diff
@@ -0,0 +1,340 @@
+                                  LogicalType key_encoded_type, LogicalType value_encoded_type, long memory_limit);
+
+    // dictionary id -> DictionaryCache
+    std::unordered_map<DictionaryId, DictionaryCachePtr> _dict_cache;
```
just cache the latest version in _dict_cache_versions.

fzhedu commented on fe/fe-core/src/main/java/com/starrocks/analysis/DictionaryExpr.java:
```diff
@@ -0,0 +1,88 @@
+    private TDictionaryExpr dictionaryExpr;
+    private boolean analyzeOnly;
+
+    public DictionaryExpr(List<Expr> params) throws SemanticException {
```
?

srlch replied:
done

fzhedu commented on fe/fe-core/src/main/java/com/starrocks/analysis/DictionaryExpr.java:
```diff
@@ -0,0 +1,88 @@
+
+    protected DictionaryExpr(DictionaryExpr other) {
+        super(other);
+        this.dictionaryExpr = other.getDictionaryExpr();
```
all parameters need to be set

srlch replied:
done

Seaven commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictionaryOperator.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+public class DictionaryOperator extends CallOperator {
```
meaningless inheritance....

srlch replied:
done

Seaven commented on fe/fe-core/src/main/java/com/starrocks/analysis/DictionaryExpr.java:
```diff
@@ -0,0 +1,88 @@
+
+import java.util.List;
+
+public class DictionaryExpr extends FunctionCallExpr {
```
meaningless inheritance....

srlch replied:
done

fzhedu commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -1847,6 +1850,128 @@ public Void visitDictQueryExpr(DictQueryExpr node, Scope context) {
+                }
+            }
+
+            final TDictionaryExpr dictionaryExpr = new TDictionaryExpr();
+            dictionaryExpr.setDict_id(dictionary.getDictionaryId());
+            dictionaryExpr.setTxn_id(GlobalStateMgr.getCurrentState().getDictionaryMgr().
+                    getLastSuccessTxnId(dictionary.getDictionaryId()));
+            dictionaryExpr.setKey_size(dictionary.getKeys().size());
+            
+            ArrayList<StructField> structFields = new ArrayList<>(valuesColumn.size());
+            for (Column column : valuesColumn) {
+                String fieldName = column.getName();
+                Type fieldType = column.getType();
+                structFields.add(new StructField(fieldName, fieldType));
+            }
+            StructType returnType = new StructType(structFields);
+            node.setType(returnType);
+            Function fn = new Function(FunctionName.createFnName(FunctionSet.DICTIONARY_GET), paramType, returnType, false);
+            fn.setBinaryType(TFunctionBinaryType.BUILTIN);
+            node.setFn(fn);
```
use to thrift?

srlch replied:
done

fzhedu commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictionaryOperator.java:
```diff
@@ -0,0 +1,48 @@
+        super(FunctionSet.DICTIONARY_GET, returnType, arguments);
+        this.dictionaryExpr = dictionaryExpr;
+        this.fn = fn;
+    }
```
clone()?

srlch replied:
done

srlch committed:
[Feature] Support Dictionary (#35559)

Why I'm doing:
StarRocks support dict_mapping function(#20961) for query a dictionary table using expression.
But there are still some limitations:
1. Only support primary key table.
2. Dictionary table lookups still require io and network overhead.
3. Multiple columns returns are not supported

What I'm doing:
StarRocks introduces dictionary objects. The dictionary object is similar to a memory table organized
into a hash table. The dictionary object caches all the data of the dictionary table to each be node,
and implements efficient dictionary queries in memory through the dictionary_get function.
Dictionary ojbect support all kinds of table modes as well as VIEW and MV.
Dictionary oject return multiple columns using struct type.

Usage:
MySQL [test]> SELECT * FROM t;
+------+------+------+
| id   | name | job1 |
+------+------+------+
|    1 |    2 |    3 |
+------+------+------+
1 row in set (0.01 sec)

MySQL [test]> CREATE DICTIONARY my_dictionary USING t (id KEY, name VALUE, job1 VALUE);
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> REFRESH DICTIONARY my_dictionary;
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> SELECT dictionary_get(my_dictionary, id) FROM t;
+-------------------------------------+
| dictionary_get('my_dictionary', id) |
+-------------------------------------+
| {name:2,job1:3}                 |
+-------------------------------------+
1 row in set (0.02 sec)

Signed-off-by: srlch <linzichao@starrocks.com>

trueeyu commented on fe/fe-core/src/main/java/com/starrocks/rpc/PBackendService.java:
```diff
@@ -90,5 +98,17 @@ public interface PBackendService {
     @ProtobufRPC(serviceName = "PBackendService", methodName = "exec_short_circuit",
```
PBackendService will be obsolete，use PInternalService instead

srlch replied:
done

trueeyu commented on be/src/exec/dictionary_cache_writer.cpp:
```diff
@@ -0,0 +1,334 @@
+            _buffer_chunk = chunk->clone_empty_with_slot();
+        }
+        DCHECK(_buffer_chunk != nullptr);
+        _buffer_chunk->append(*chunk.get());
```
*chunk.get() ---> *chunk

srlch replied:
done

trueeyu commented on be/src/exec/dictionary_cache_writer.cpp:
```diff
@@ -0,0 +1,334 @@
+        }
+
+        st = _wait_response(closures);
+        break;
```
always break?

srlch replied:
Yes, it always break. `While(True);break;`  is a way of implementing goto statements. It can do something. When there is an error in the middle of the code process, it can be jumped out in advance and error handling can be carried out directly

srlch committed:
Fix be service problem

Signed-off-by: srlch <linzichao@starrocks.com>

fzhedu commented on gensrc/proto/internal_service.proto:
```diff
@@ -621,5 +671,12 @@ service PInternalService {
+
     rpc exec_short_circuit(PExecShortCircuitRequest) returns (PExecShortCircuitResult);
+
+    rpc refresh_dictionary_cache_begin(PRefreshDictionaryCacheBeginRequest) returns (PRefreshDictionaryCacheBeginResult);
+    rpc refresh_dictionary_cache_commit(PRefreshDictionaryCacheCommitRequest) returns (PRefreshDictionaryCacheCommitResult);
+
+    rpc clear_dictionary_cache(starrocks.PClearDictionaryCacheRequest) returns (starrocks.PClearDictionaryCacheResult);
```
can merge this to one RPC, like rpc process_dictionary_cache(request, cmd), as they use the same results only contains a status.

srlch replied:
fix it in next pr

fzhedu commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictionaryExprOperator.java:
```diff
@@ -0,0 +1,130 @@
+import java.util.List;
+import java.util.Objects;
+
+public class DictionaryExprOperator extends ScalarOperator {
```
Expr and Operator are different, rename to DictonaryOperator? or others

srlch replied:
fix it in next pr

srlch commented on be/src/exprs/dictionary_expr.cpp:
```diff
@@ -0,0 +1,132 @@
+    std::vector<ColumnId> value_cids(_schema->fields().size() - _dictionary_expr.key_size);
+    std::iota(value_cids.begin(), value_cids.end(), _dictionary_expr.key_size);
+    _value_chunk = ChunkHelper::new_chunk(Schema(_schema.get(), value_cids), 0);
+    _value_chunk->reset();
```
fix it in next pr

trueeyu commented on be/src/service/internal_service.cpp:
```diff
@@ -775,6 +776,84 @@ void PInternalServiceImplBase<T>::get_file_schema(google::protobuf::RpcControlle
+    auto st = StorageEngine::instance()->dictionary_cache_manager()->begin(dict_id, txn_id);
+    if (!st.ok()) {
+        LOG(WARNING) << st.get_error_msg();
+        Status::InternalError(st.get_error_msg()).to_protobuf(response->mutable_status());
```
return ?

srlch replied:
Fix in next pr

trueeyu commented on be/src/service/internal_service.cpp:
```diff
@@ -775,6 +776,84 @@ void PInternalServiceImplBase<T>::get_file_schema(google::protobuf::RpcControlle
+    auto st = StorageEngine::instance()->dictionary_cache_manager()->commit(dict_id, txn_id);
+    if (!st.ok()) {
+        LOG(WARNING) << st.get_error_msg();
+        Status::InternalError(st.get_error_msg()).to_protobuf(response->mutable_status());
```
return ?

srlch replied:
Fix in next pr

trueeyu commented on be/src/storage/dictionary_cache_manager.cpp:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    // release after get the encoded column
+    key_chunk->reset();
```
why reset()?

srlch replied:
release memory as soon as possible

srlch replied:
Fix in next pr

trueeyu commented on be/src/runtime/runtime_state.cpp:
```diff
@@ -54,6 +54,7 @@
 #include "runtime/mem_tracker.h"
 #include "runtime/query_statistics.h"
 #include "runtime/runtime_filter_worker.h"
+#include "storage/storage_engine.h"
```
remove storage_engine.h?

srlch replied:
Fix in next pr

trueeyu commented on be/src/runtime/dictionary_cache_sink.cpp:
```diff
@@ -0,0 +1,39 @@
+namespace starrocks {
+
+Status DictionaryCacheSink::init(const TDataSink& t_sink, RuntimeState* state) {
+    RETURN_IF_ERROR(DataSink::init(t_sink, state));
```
return DataSink::init(t_sink, state);

srlch replied:
It seems that it is standard  implementation for sink?

trueeyu commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictionaryExprOperator.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public boolean isNullable() {
+        return false;
```
false?

srlch replied:
Always false by definition

trueeyu commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictionaryExprOperator.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public String toString() {
+        return "DICTIONAR_GET";
```
DICTIONAR_GET or DICTIONAR ?

srlch replied:
typo problem, fix in next pr

trueeyu commented on fe/fe-core/src/main/java/com/starrocks/sql/ast/DictionaryExpr.java:
```diff
@@ -0,0 +1,132 @@
+
+import java.util.List;
+
+public class DictionaryExpr extends Expr {
```
DictionaryGetExpr ?

srlch replied:
Fix this name in next pr
